runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingEngineWorkCommitterTest.java

```diff
@@ -0,0 +1,208 @@
+
+    Map<Long, WorkItemCommitRequest> committed =
+        fakeWindmillServer.waitForAndGetCommits(commits.size());
+
+    for (Commit commit : commits) {
+      WorkItemCommitRequest request = committed.get(commit.work().getWorkItem().getWorkToken());
+      assertNotNull(request);
+      assertThat(request).isEqualTo(commit.request());
+    }
+  }
+
+  @Test
+  public void testCommit_handlesFailedCommits() {
+    Set<CompleteCommit> failedCommits = new HashSet<>();
+    workCommitter = createWorkCommitter(failedCommits::add);
+    List<Commit> commits = new ArrayList<>();
+    for (int i = 1; i <= 10; i++) {
+      Work work = createMockWork(i, ignored -> {});
+      // Fail half of the work.
+      if (i % 2 == 0) {
+        work.setFailed();
+      }
+      WorkItemCommitRequest commitRequest =
+          WorkItemCommitRequest.newBuilder()
+              .setKey(work.getWorkItem().getKey())
+              .setShardingKey(work.getWorkItem().getShardingKey())
+              .setWorkToken(work.getWorkItem().getWorkToken())
+              .setCacheToken(work.getWorkItem().getCacheToken())
+              .build();
+      commits.add(Commit.create(commitRequest, createComputationState("computationId-" + i), work));
+    }
+
+    workCommitter.start();
+    commits.forEach(workCommitter::commit);
+
+    Map<Long, WorkItemCommitRequest> committed =
+        fakeWindmillServer.waitForAndGetCommits(commits.size() / 2);
+
+    for (Commit commit : commits) {
+      if (commit.work().isFailed()) {
+        assertThat(failedCommits).contains(asCompleteCommit(commit));
+        assertThat(committed).doesNotContainKey(commit.work().getWorkItem().getWorkToken());
+      }
+    }
+  }
+
+  @Test
+  public void testCommit_handlesCompleteCommits() {
+    Set<CompleteCommit> completeCommits = new HashSet<>();
+    workCommitter = createWorkCommitter(completeCommits::add);
+    List<Commit> commits = new ArrayList<>();
+    for (int i = 1; i <= 5; i++) {
+      Work work = createMockWork(i, ignored -> {});
+      WorkItemCommitRequest commitRequest =
+          WorkItemCommitRequest.newBuilder()
+              .setKey(work.getWorkItem().getKey())
+              .setShardingKey(work.getWorkItem().getShardingKey())
+              .setWorkToken(work.getWorkItem().getWorkToken())
+              .setCacheToken(work.getWorkItem().getCacheToken())
+              .build();
+      commits.add(Commit.create(commitRequest, createComputationState("computationId-" + i), work));
+    }
+
+    workCommitter.start();
+    commits.forEach(workCommitter::commit);
+
+    Map<Long, WorkItemCommitRequest> committed =
+        fakeWindmillServer.waitForAndGetCommits(commits.size());
+
+    for (Commit commit : commits) {
+      WorkItemCommitRequest request = committed.get(commit.work().getWorkItem().getWorkToken());
+      assertNotNull(request);
+      assertThat(request).isEqualTo(commit.request());
+      assertThat(completeCommits).contains(asCompleteCommit(commit));
```

Comment:
add a test that verifies some particular keys failing in windmill, that the non-ok commit status is plumbed correctly

Reply:
done had to make some changes to FakeWindmillServer